File: LUA/28-regex-patterns-and-intro-to-databases.lua
local N = io.stdin:read("*n", "*l")

names = {}

for Nitr = 1, N do
    local firstNameEmailID = {}
    for token in string.gmatch(io.stdin:read("*l"), "[^%@]+") do
        table.insert(firstNameEmailID, token)
    end -- for

    local firstName = firstNameEmailID[1]
    local emailID = firstNameEmailID[2]

    if emailID == "gmail.com" then
        table.insert(names, firstName:match("%w+"))
    end -- if
end -- for

table.sort(names)

for i = 1, #names do
    print(names[i])
end --for

File: JAVA/28-regex-patterns-and-intro-to-databases.java
import java.util.Collections;
import java.util.LinkedList;
import java.util.Scanner;

class RegExp
{
  private static final Scanner scanner = new Scanner(System.in);
  
  public static void main(String[] args)
  {
    LinkedList<String> names = new LinkedList<>();
    int                N     = scanner.nextInt();
    scanner.skip("(\r\n|[\n\r\u2028\u2029\u0085])?");
    for (int NItr = 0; NItr < N; NItr++)
    {
      String[] firstNameEmailID = scanner.nextLine().split("\\s[a-zA-Z.]*");
      String   firstName        = firstNameEmailID[0];
      String   emailID          = firstNameEmailID[1];
      
      if (emailID.equals("@gmail.com"))
        names.add(firstName);
    }// end for - NItr
    
    Collections.sort(names);
    
    for (String s : names)
      System.out.println(s);
    
    scanner.close();
  }// end main
}// end RegExp - class